pandas/tests/sparse/test_combine_concat.py

```diff
@@ -317,37 +318,43 @@ def test_concat_axis1(self):
         assert isinstance(res, pd.SparseDataFrame)
         tm.assert_frame_equal(res.to_dense(), exp)
 
-    def test_concat_sparse_dense(self):
-        sparse = self.dense1.to_sparse()
-
+    @pytest.mark.parametrize('fill_value', [None, 0])
+    def test_concat_sparse_dense(self, fill_value):
+        sparse = self.dense1.to_sparse(fill_value=fill_value)
         res = pd.concat([sparse, self.dense2])
         exp = pd.concat([self.dense1, self.dense2])
-        assert isinstance(res, pd.SparseDataFrame)
-        tm.assert_frame_equal(res.to_dense(), exp)
 
-        res = pd.concat([self.dense2, sparse])
-        exp = pd.concat([self.dense2, self.dense1])
-        assert isinstance(res, pd.SparseDataFrame)
-        tm.assert_frame_equal(res.to_dense(), exp)
-
-        sparse = self.dense1.to_sparse(fill_value=0)
-
-        res = pd.concat([sparse, self.dense2])
-        exp = pd.concat([self.dense1, self.dense2])
         assert isinstance(res, pd.SparseDataFrame)
         tm.assert_frame_equal(res.to_dense(), exp)
 
         res = pd.concat([self.dense2, sparse])
         exp = pd.concat([self.dense2, self.dense1])
+
         assert isinstance(res, pd.SparseDataFrame)
         tm.assert_frame_equal(res.to_dense(), exp)
 
         res = pd.concat([self.dense3, sparse], axis=1)
         exp = pd.concat([self.dense3, self.dense1], axis=1)
-        assert isinstance(res, pd.SparseDataFrame)
+        # See GH18914 and #18686 for why this should be
+        # A DataFrame
+        assert type(res) is pd.DataFrame
+        # See GH16874
+        assert not res.isnull().empty
+        assert not res[res.columns[0]].empty
+        assert res.iloc[0, 0] == self.dense3.iloc[0, 0]
+
+        for column in self.dense3.columns:
+            tm.assert_series_equal(res[column], exp[column])
```

Comment:
I would be all for splitting this long test up a bit

Reply:
Alright split it up a bit. Also added some new test cases since it wasn't covering all aspects like concating onself.